src/main/java/de/unijena/cheminf/fragment/fingerprint/FragmentFingerprinter.java

```diff
@@ -0,0 +1,464 @@
+    public IBitFingerprint getBitFingerprint(List<String> aListOfUniqueSmiles) throws NullPointerException, IllegalArgumentException {
+        Objects.requireNonNull(aListOfUniqueSmiles, "aFragmentList (list of string instances) is null.");
+        BitSet tmpBitSet = new BitSet(this.fragmentArray.length);
+        for (String tmpUniqueSmiles : aListOfUniqueSmiles) {
+            Objects.requireNonNull(tmpUniqueSmiles, "aFragmentList (at least one list element) is null.");
+            if(tmpUniqueSmiles.isBlank() ||tmpUniqueSmiles.isEmpty()) {
+                throw new IllegalArgumentException("aFragmentList (at least one list element) is blank/empty.");
+            }
+        }
+        Set<String> tmpSet = new HashSet<>(aListOfUniqueSmiles);
+        ArrayList<String> tmpUniqueSmiles = new ArrayList<>(tmpSet);
+        this.copyListBit = (ArrayList<String>) tmpUniqueSmiles.clone();
+        for(String tmpUniqueSmilesWithoutDuplicates : tmpUniqueSmiles) {
+            if (this.uniqueSmilesToPositionMap.containsKey(tmpUniqueSmilesWithoutDuplicates)) {
+                int tmpPosition = uniqueSmilesToPositionMap.get(tmpUniqueSmilesWithoutDuplicates);
+                tmpBitSet.set(tmpPosition, true);
+            }
+        }
+        this.cacheBitFingerprint = new BitSetFingerprint(tmpBitSet);
+        return this.cacheBitFingerprint;
+    }
+    //
+    /**
+     * Method to generate count fingerprint.
+     * The input map assigns the unique SMILES to their frequencies. The frequency of these unique SMILES are compared
+     * with the predefined fragments.
+     * If there is a match between the unique SMILES,
+     * the corresponding position of the unique SMILES is determined from the "uniqueSmilesToPositionMap" and
+     * the corresponding frequency from "aUniqueSmilesToFrequencyMap". The new "rawCountMap" then
+     * maps the position of the unique SMILES to the corresponding frequency. If the specified unique SMILES
+     * are not part of the predefined fragments (unique SMILES), they are ignored.
+     *
+     * @param aUniqueSmilesToFrequencyMap  is a map that maps fragments in the form of unique SMILES to the
+     * frequency of unique SMILES.
+     * To be able to calculate the fingerprint for a molecule, the fragments must belong to a molecule.
+     * @return count fingerprint
+     * @throws NullPointerException  is thrown if the map aUniqueSmilesToFrequencyMap is
+     * null or contains keys or values that are null respectively.
+     * @throws IllegalArgumentException is thrown if the map aUniqueSmilesToFrequencyMap
+     * contains keys or values that are blank/empty, respectively.
+     */
+    @Override
+    public ICountFingerprint getCountFingerprint(Map<String, Integer> aUniqueSmilesToFrequencyMap) throws NullPointerException,IllegalArgumentException {
+        this.cacheRawCountMap = new HashMap<>(this.fragmentArray.length);
+        this.copyListCount = new ArrayList<>(aUniqueSmilesToFrequencyMap.size());
```

Comment:
The "loading factor thing" also applies to ArrayLists I think.

Reply:
"Each ArrayList instance has a capacity. The capacity is the size of the array used to store the elements in the list. It is always at least as large as the list size. As elements are added to an ArrayList, its capacity grows automatically. The details of the growth policy are not specified beyond the fact that adding an element has constant amortized time cost.

An application can increase the capacity of an ArrayList instance before adding a large number of elements using the ensureCapacity operation. This may reduce the amount of incremental reallocation."

Does not seem like ArrayLists have load factors, but its internal resizing behaviour is not really specified here.
(source: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ArrayList.html)

Reply:
https://www.geeksforgeeks.org/how-to-increase-the-capacity-size-of-arraylist/
I guess if you set the initial size as high as needed here, you are on the safe side. Seems to me like this ensureCapacity() method is meant for cases where you didn't know the needed size yet when creating the list.

Reply:
Just set it high enough, I'll do a bit of research